google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/BitReversedSequenceStyleGenerator.java

```diff
@@ -0,0 +1,169 @@
+      this.initialValue = initialValue;
+    }
+
+    @Override
+    public void registerExportables(Database database) {
+      super.registerExportables(database);
+      // Replace the init command for the table-backed sequence.
+      // Hibernate by default generates an 'insert into table_name values (?)' statement.
+      // That is not supported by Cloud Spanner, as Cloud Spanner requires the insert statement to
+      // include the column name(s) that are being used in the insert statement.
+      final Namespace namespace =
+          database.locateNamespace(qualifiedName.getCatalogName(), qualifiedName.getSchemaName());
+      Table table = namespace.locateTable(qualifiedName.getObjectName());
+      if (table != null) {
+        Dialect dialect = jdbcEnvironment.getDialect();
+        String valueColumnNameText = valueColumnNameIdentifier.render(dialect);
+        table.addInitCommand(
+            context ->
+                new ReplaceInitCommand(
+                    "insert into "
+                        + context.format(table.getQualifiedTableName())
+                        + " ("
+                        + valueColumnNameText
+                        + ") values ( "
+                        + initialValue
+                        + " )"));
+      }
+    }
+  }
+
+  /**
+   * Acts as a replacement for other {@link InitCommand}.
+   *
+   * <ol>
+   *   <li>If the list contains at least one {@link ReplaceInitCommand} and at least one normal
+   *       {@link InitCommand}, then all normal {@link InitCommand}s will be ignored during
+   *       execution and only the {@link ReplaceInitCommand}s will be executed.
+   *   <li>If the list only contains {@link ReplaceInitCommand}s, nothing will be executed.
+   *   <li>If the list only contains normal {@link InitCommand}s, all normal {@link InitCommand}s
+   *       will be executed as normal.
+   * </ol>
+   */
+  public static class ReplaceInitCommand extends InitCommand {
+    public ReplaceInitCommand(String... initCommands) {
+      super(initCommands);
+    }
+  }
+
+  @Override
+  protected DatabaseStructure buildTableStructure(
+      Type type,
+      Properties params,
+      JdbcEnvironment jdbcEnvironment,
+      QualifiedName sequenceName,
+      int initialValue,
+      int incrementSize) {
+    final Identifier valueColumnName = determineValueColumnName(params, jdbcEnvironment);
+    return new SpannerSequenceTableStructure(
+        jdbcEnvironment,
+        sequenceName,
+        valueColumnName,
+        initialValue,
+        incrementSize,
+        type.getReturnedClass());
+  }
+
+  /**
+   * Generates a new ID. This uses the normal sequence strategy, but the returned ID is bit-reversed
+   * before it is returned to the application.
+   */
+  @Override
+  public Serializable generate(SharedSessionContractImplementor session, Object object)
+      throws HibernateException {
+    Serializable id = super.generate(session, object);
```

Comment:
From what value is the sequence initiated when the application starts up ?

Reply:
Depending on the initial state of the underlying database:
1. The sequence (table) does not yet exist: A new table is created and a row is inserted with the initial value that is given here: https://github.com/GoogleCloudPlatform/google-cloud-spanner-hibernate/blob/e17095bdaf584c9f471043b870722944a6ac74d5/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/BitReversedSequenceStyleGenerator.java#L60
2. The sequence (table) exists (or was just created in step 1): The current value in the table.

So in the example given in the code sample, the sequence (table) will be created with an initial value of 50,000. This value will increase as more and more IDs are generated. The application will read the current value from this table at startup, and when it has exhausted its current range.